prog/hints.go

```diff
@@ -104,10 +104,18 @@ func generateHints(compMap CompMap, arg Arg, exec func()) {
 }
 
 func checkConstArg(arg *ConstArg, compMap CompMap, exec func()) {
-	original := arg.Val
+	argSize := arg.Size()
+	if argSize > 8 {
+		argSize = 8
+	}
+	limit := getSizeLimit(argSize)
+	original := arg.Val & limit
 	// Note: because shrinkExpand returns a map, order of programs is non-deterministic.
 	// This can affect test coverage reports.
 	for _, replacer := range shrinkExpand(original, compMap) {
+		if replacer > limit {
```

Comment:
shrinkExpand produces set of valid replacements. It rejects lots of invalid combinations already. This function does not do the argument selection logic, it has different responsibilities.
All logic related to argument selection/rejection need to go into shrinkExpand. It also can be unit-tested.